### PR TITLE
Adding transformers to Encounter Components

### PR DESCRIPTION
**Pull request by ElliotB-96**

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java [hunk 1]
```diff
@@ -48,22 +52,29 @@
...
+        .put(ResourceType.DocumentReference, this::mapDocumentReference)
+        .put(ResourceType.Immunization, this::mapImmunization)
         .put(ResourceType.List, this::mapListResource)
-        .put(ResourceType.ProcedureRequest, this::mapProcedureRequest)
+        .put(ResourceType.MedicationRequest, this::mapMedicationRequest)
         .put(ResourceType.Observation, this::mapObservation)
-        .put(ResourceType.Immunization, this::mapImmunization)
-        .put(ResourceType.Condition, this::mapCondition)
+        .put(ResourceType.ProcedureRequest, this::mapProcedureRequest)
+        .put(ResourceType.ReferralRequest, this::mapReferralRequest)
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:63]: This whole thing might be cleaner with `Map.of()`
```
    private final Map<ResourceType, Function<Resource, String>> encounterComponents = Map.of(
        ResourceType.AllergyIntolerance, this::mapAllergyIntolerance,
        ResourceType.Condition, this::mapCondition,
        ResourceType.DocumentReference, this::mapDocumentReference,
        ResourceType.Immunization, this::mapImmunization,
        ResourceType.List, this::mapListResource,
        ResourceType.MedicationRequest, this::mapMedicationRequest,
        ResourceType.Observation, this::mapObservation,
        ResourceType.ProcedureRequest, this::mapProcedureRequest,
        ResourceType.ReferralRequest, this::mapReferralRequest);
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java [hunk 2]
```diff
@@ -96,12 +107,34 @@ private String mapDefaultNotImplemented(Resource resource) {
...
+    }
+
+    private String mapImmunization(Resource resource) {
+        return immunizationObservationStatementMapper.mapImmunizationToObservationStatement((Immunization) resource, IS_NESTED);
+    }
+
+    private String mapListResource(Resource resource) {
+        ListResource listResource = (ListResource) resource;
+
+        if (listResource.hasEntry() && hasCode(listResource.getCode(), COMPONENTS_LISTS)) {
```
Comment (andrewwa-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:129]: Are there lists of other codes that we're not yet mapping, or would another list indicate garbage in?

Comment (mattd-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:126]: There are the only lists we're concerned with. It is the consultation/topic/heading list for structured consultations which are flattened for EHR output.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java [hunk 3]
```diff
@@ -46,24 +48,29 @@
...
+    private final Map<ResourceType, Function<Resource, String>> encounterComponents = Map.of(
+        ResourceType.AllergyIntolerance, this::mapAllergyIntolerance,
+        ResourceType.Condition, this::mapCondition,
+        ResourceType.DocumentReference, this::mapDocumentReference,
+        ResourceType.Immunization, this::mapImmunization,
+        ResourceType.List, this::mapListResource,
+        ResourceType.MedicationRequest, this::mapMedicationRequest,
+        ResourceType.Observation, this::mapObservation,
+        ResourceType.ProcedureRequest, this::mapProcedureRequest,
+        ResourceType.ReferralRequest, this::mapReferralRequest);
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:60]: I think we can begin to phase out our placeholder now

```suggestion
        ResourceType.ReferralRequest, this::mapReferralRequest,
        ResourceType.DiagnosticReport, this:: mapDefaultNotImplemented);
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java [hunk 4]
```diff
@@ -85,6 +92,7 @@ private String mapListResourceToComponents(ListResource listReferencedToEncounte
 
     private Optional<String> mapItemToComponent(ListResource.ListEntryComponent item) {
         Optional<Resource> resource = messageContext.getInputBundleHolder().getResource(item.getItem().getReferenceElement());
+        resource.map(Resource::getResourceType).ifPresent(System.out::println);
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:95]: remove this println

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java [hunk 5]
```diff
@@ -85,6 +92,7 @@ private String mapListResourceToComponents(ListResource listReferencedToEncounte
...
 
         return resource.map(value -> encounterComponents.getOrDefault(value.getResourceType(), this::mapDefaultNotImplemented)
```
Comment (mattd-kainos) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:97]: As above about phasing out our placeholder

```suggestion
        return resource.map(value -> encounterComponents.get(value.getResourceType()).apply(value));
```

Comment (mattd-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/EncounterComponentsMapper.java:97]: Nevermind, I'll do this in the UAT ticket

**Review by mattd-kainos — APPROVED** [review]

**Review by andrewwa-kainos — APPROVED** [review]

**Pull request merged**